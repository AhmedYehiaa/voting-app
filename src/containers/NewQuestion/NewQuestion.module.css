.newQuestionContainer {
  width: 50%;
  margin: auto;
  padding: 50px;
}

.newQuestionForm {
  display: flex;
  flex-direction: column;
  align-items: center;
  background: #f3f3f3;
}

.newQuestionForm ul {
  list-style: none;
}

.choiceInput {
  display: flex;
  flex-direction: column;
  /* align-items: center; */
  /* background: #ccc; */
}
.choicesListWrapper {
  width: 80%;
  margin: 10px auto;
  background-color: #f7fafc;
}

.choicesListWrapper ul {
  padding: 0;
  margin: 0 auto;
}
.choicesList {
  width: 70%;
}
.choicesList li {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.questionInput,
.choiceInput {
  width: 90%;
  padding: 30px 0;
}

.questionInput input,
.choiceInput input {
  width: 80%;
  padding: 10px;
  border: none;
}
.inputGroup {
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding-top: 10px;
  font-size: 16px;
}

.input {
  font-size: 16px;
  border-radius: 5px;
  outline: none;
}

.addChoiceBtn {
  width: 110px;
  border: none;
  margin-left: 10px;
  font-weight: bold;
  outline: none;
  border-radius: 5px;
  background: rgb(0, 41, 55);
  color: #fff;
}
.addChoiceBtn:hover {
  cursor: pointer;
}

.removeBtn {
  padding: 5px;
  border-radius: 5px;
  border: none;
  font-size: 14px;
  background-color: #e2e8f0;
}
.removeBtn:hover {
  cursor: pointer;
}
.submitButton {
  width: 50%;
  padding: 10px;
  margin: 10px auto;
  border: none;
  border-radius: 5px;
  font-weight: bold;
  text-transform: uppercase;
  background: rgb(0, 41, 55);
  color: #fff;
}

.submitButton:hover {
  cursor: pointer;
}

.submitButton:disabled,
.submitButton[disabled] {
  border: 1px solid #999999;
  background-color: #cccccc;
  color: #2d3748;
}
.submitButton[disabled]:hover {
  cursor: not-allowed;
}

@media only screen and (max-width: 900px) {
  .questionInput,
  .choiceInput {
    width: 90%;
  }
  .newQuestionContainer {
    width: 90%;
    margin: auto;
    padding: 20px 0;
  }
}
