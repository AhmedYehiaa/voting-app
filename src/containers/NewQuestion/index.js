import React, { useState } from "react";
import { withRouter } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";

import styles from "./NewQuestion.module.css";
import Spinner from "../../components/Spinner";

import { addNewQuestion } from "../../services/APIcalls";

const NewQuestion = ({ history }) => {
  const [loading, setLoading] = useState(false);
  const [question, setQuestion] = useState("");
  const [choices, setChoices] = useState([]);
  const [newChoice, setNewChoice] = useState("");

  const addChoice = () => {
    if (newChoice.trim()) {
      setChoices([...choices, newChoice]);
      setNewChoice("");
    }
  };

  const handleQuestionChange = event => {
    event.preventDefault();
    const { value } = event.target;
    setQuestion(value);
  };

  const handleChoiceChange = event => {
    event.preventDefault();
    const { value } = event.target;
    setNewChoice(value);
  };

  const handleDelete = index => {
    const filteredChoices = [...choices];
    filteredChoices.splice(index, 1);
    setChoices(filteredChoices);
  };

  const handleSubmit = async event => {
    event.preventDefault();
    const body = {
      question,
      choices
    };
    try {
      setLoading(true);
      await addNewQuestion(body);
      setLoading(false);
      history.replace("/");
    } catch (error) {
      setLoading(false);
      history.replace("/not-found");
    }
  };

  const rendeCondition = () => {
    if (loading) return <Spinner />;
    else {
      return (
        <div className={styles.newQuestionContainer}>
          <div className={styles.newQuestionForm}>
            <div className={styles.questionInput}>
              <div className={styles.inputGroup}>
                <input
                  value={question}
                  className={styles.input}
                  placeholder="Add question"
                  onChange={handleQuestionChange}
                />
              </div>
            </div>
            <div className={styles.choiceInput}>
              <div className={styles.inputGroup}>
                <input
                  type="text"
                  id="currentChoice"
                  name="currentChoice"
                  className={styles.input}
                  value={newChoice}
                  placeholder="Add choice"
                  onChange={handleChoiceChange}
                />
                <button
                  onClick={addChoice}
                  className={styles.addChoiceBtn}
                  data-testid="addChoiceBtn"
                >
                  Add choice <FontAwesomeIcon icon={faPlus} />
                </button>
              </div>
            </div>
            {choices.length > 0 ? (
              <div className={styles.choicesListWrapper}>
                <ul className={styles.choicesList} data-testid="choicesList">
                  {choices.map((choice, index) => (
                    <li key={choice + index}>
                      <p data-testid="choice">
                        {index + 1}. {"   "}
                        {choice}
                      </p>
                      <button
                        className={styles.removeBtn}
                        data-testid={`removeChoiceBtn-${index + 1}`}
                        title="remove"
                        onClick={() => handleDelete(index)}
                      >
                        Remove
                      </button>
                    </li>
                  ))}
                </ul>
              </div>
            ) : null}
            <button
              disabled={!(question && choices.length >= 2)}
              className={styles.submitButton}
              onClick={handleSubmit}
            >
              Submit
            </button>
          </div>
        </div>
      );
    }
  };

  return rendeCondition();
};

export default withRouter(NewQuestion);
