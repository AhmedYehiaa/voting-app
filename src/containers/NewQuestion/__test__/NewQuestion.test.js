import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter as Router } from "react-router-dom";
import { render, within, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

import NewQuestion from "../index";

describe("<NewQuestion />", () => {
  it("render without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Router>
        <NewQuestion />
      </Router>,
      div
    );
  });

  it("add new question with choices", () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <Router>
        <NewQuestion />
      </Router>
    );
    const questionInput = getByPlaceholderText("Add question");
    const choiceInput = getByPlaceholderText("Add choice");
    const addChoiceBtn = getByTestId("addChoiceBtn");
    const questionTitle = "new question ?";
    const newChoice = "new choice";

    fireEvent.change(questionInput, { target: { value: questionTitle } });
    fireEvent.change(choiceInput, { target: { value: newChoice } });
    fireEvent.click(addChoiceBtn);

    expect(choiceInput).toHaveValue("");
    expect(getByTestId("choice").textContent).toBe(`1.    ${newChoice}`);
    fireEvent.change(choiceInput, { target: { value: newChoice } });
  });

  it("remove choice from choices list", () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <Router>
        <NewQuestion />
      </Router>
    );
    const questionInput = getByPlaceholderText("Add question");
    const choiceInput = getByPlaceholderText("Add choice");
    const addChoiceBtn = getByTestId("addChoiceBtn");
    const questionTitle = "new question ?";
    const newChoice = "new choice";

    fireEvent.change(questionInput, { target: { value: questionTitle } });
    fireEvent.change(choiceInput, { target: { value: newChoice } });
    fireEvent.click(addChoiceBtn);

    expect(choiceInput).toHaveValue("");

    fireEvent.change(choiceInput, { target: { value: newChoice } });
    fireEvent.click(addChoiceBtn);

    expect(getByTestId("choicesList").children.length).toBe(2);
    const removeChoicBtn = within(getByTestId("removeChoiceBtn-1")).getByText(
      /remove/i
    );
    fireEvent.click(removeChoicBtn);
    expect(getByTestId("choicesList").children.length).toBe(1);
  });

  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Router>
          <NewQuestion />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
