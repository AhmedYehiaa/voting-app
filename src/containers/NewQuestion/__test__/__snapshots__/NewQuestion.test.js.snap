// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewQuestion /> matches snapshot 1`] = `
<div
  className="newQuestionContainer"
>
  <div
    className="newQuestionForm"
  >
    <div
      className="questionInput"
    >
      <div
        className="inputGroup"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Add question"
          value=""
        />
      </div>
    </div>
    <div
      className="choiceInput"
    >
      <div
        className="inputGroup"
      >
        <input
          className="input"
          id="currentChoice"
          name="currentChoice"
          onChange={[Function]}
          placeholder="Add choice"
          type="text"
          value=""
        />
        <button
          className="addChoiceBtn"
          data-testid="addChoiceBtn"
          onClick={[Function]}
        >
          Add choice 
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-plus fa-w-14 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      className="submitButton"
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
  </div>
</div>
`;
